firedrake/utility_meshes.py

```diff
@@ -863,6 +864,116 @@ def UnitIcosahedralSphereMesh(refinement_level=0, degree=1, reorder=None,
+    :kwarg reorder: (optional), should the mesh be reordered?
+    :kwarg comm: Optional communicator to build the mesh on (defaults to
+        COMM_WORLD).
+    """
+    if refinement_level < 0 or refinement_level % 1:
+            raise RuntimeError("Number of refinements must be a non-negative integer")
+
+    if degree < 1:
+        raise ValueError("Mesh coordinate degree must be at least 1")
+    # vertices of an octahedron of radius 1
+    if hemisphere:
+        vertices = np.array([[1.0, 0.0, 0.0],
+                             [0.0, 1.0, 0.0],
+                             [0.0, 0.0, 1.0],
+                             [-1.0, 0.0, 0.0],
+                             [0.0, -1.0, 0.0]])
+        # faces of the base octahedron
+        faces = np.array([[0, 1, 2],
+                          [0, 2, 4],
+                          [1, 2, 3],
+                          [2, 3, 4]], dtype=np.int32)
+    else:
+        vertices = np.array([[1.0, 0.0, 0.0],
+                             [0.0, 1.0, 0.0],
+                             [0.0, 0.0, 1.0],
+                             [-1.0, 0.0, 0.0],
+                             [0.0, -1.0, 0.0],
+                             [0.0, 0.0, -1.0]])
+        # faces of the base octahedron
+        faces = np.array([[0, 1, 2],
+                          [0, 1, 5],
+                          [0, 2, 4],
+                          [0, 4, 5],
+                          [1, 2, 3],
+                          [1, 3, 5],
+                          [2, 3, 4],
+                          [3, 4, 5]], dtype=np.int32)
+
+    plex = mesh._from_cell_list(2, faces, vertices, comm)
+    plex.setRefinementUniform(True)
+    for i in range(refinement_level):
+        plex = plex.refine()
+
+    # build the initial mesh
+    m = mesh.Mesh(plex, dim=3, reorder=reorder)
+    if degree > 1:
+        # use it to build a higher-order mesh
+        new_coords = function.Function(functionspace.VectorFunctionSpace(m, "CG", degree))
+        x, y, z = ufl.SpatialCoordinate(m)
+        new_coords.interpolate(ufl.as_vector([x, y, z]))
+        m = mesh.Mesh(new_coords)
+
+    # remap to a cone
+    x, y, z = ufl.SpatialCoordinate(m)
+    tol = 1.0e-8
+    rnew = ufl.Max(1 - abs(z), 0)
+    x0 = x/(rnew + tol)
+    y0 = y/(rnew + tol)
+    theta = ufl.conditional(ufl.ge(y0, 0),
+                            ufl.pi/2*(1-x0),
+                            ufl.pi/2.0*(x0-1))
+    Vc = m.coordinates.function_space()
+    xcone = Function(Vc).interpolate(ufl.as_vector([ufl.cos(theta)*rnew,
+                                                    ufl.sin(theta)*rnew, z]))
+    m.coordinates.assign(xcone)
+
+    # push out to a sphere
+    phi = ufl.pi*z/2
+    rnew += tol
+    rnew2 = ufl.cos(phi) + tol
+    znew = ufl.sin(phi)
+    xsphere = Function(Vc).interpolate(ufl.as_vector([x*rnew2/rnew,
+                                                      y*rnew2/rnew, znew]))
+    m.coordinates.assign(xsphere*radius)
```

Comment:
maybe `*Constant(radius)` to avoid generating and compiling a new C file per radius